optin/puller/workflows/payout.py

```diff
@@ -14,48 +16,74 @@
 # example for processing and paying out post optins
 
 # TODO_:
-# 1. add check_cosigners to MultisigDatabase
 # 2. implement send_funds
-# 3. check finalized and update completed db
 
 
 def send_funds(_):
 	# TODO_: implement this
 	return Hash256('')
 
 
+class Databases:
+	def __init__(self, completed_connection, inprogress_connection, multisig_connection):
+		self.completed = CompletedOptinDatabase(completed_connection)
+		self.inprogress = InProgressOptinDatabase(inprogress_connection)
+		self.multisig = MultisigDatabase(multisig_connection)
+
+
+async def process_sent_requests(sent_requests, databases, client, symbol_network):
+	finalized_transaction_hashes = set(await filter_finalized_transactions(client, set(
+		request.transaction_hash for request in request_group for request_group in sent_requests.values()
+	)))
+	for address, request_group in sent_requests.items():
+		for request in request_group:
+			if request.transaction_hash in finalized_transaction_hashes:
+				databases.inprogress.set_request_status(request, OptinRequestStatus.COMPLETED, request.transaction_hash)
+
+		# TODO_: lookup balance
+		balance = 0
+		databases.completed.insert_mapping(
+			{str(address): balance},
+			{str(symbol_network.public_key_to_address(request_group.requests[0].destination_public_key)): balance})
+
+
+async def process_unprocessed_requests(unprocessed_requests, sent_requests, databases):
+	for address, request_group in unprocessed_requests.items():
+		if request_group.is_error:
+			# inconsistent error detection - what to do?
+			continue
+
+		if address in sent_requests or databases.completed.is_opted_in(address):
+			databases.inprogress.set_request_status(address, OptinRequestStatus.DUPLICATE, None)
+		else:
+			if databases.multisig.check_cosigners(address, [request.address for request in request_group.requests]):
+				sent_transaction_hash = send_funds(request_group.requests[0].destination_public_key)
```

Comment:
one thing is not clear to me (bit unrelated to this review), where do we send _non_ multisig payouts?

Reply:
here ^^